firedrake/slate/slac/kernel_builder.py

```diff
@@ -36,12 +40,25 @@ def __init__(self, expression, tsfc_parameters=None):
         self.needs_mesh_layers = False
         self.oriented = False
         self.finalized_ast = None
+        self._is_finalized = False
 
         # Generate coefficient map (both mixed and non-mixed cases handled)
         self.coefficient_map = prepare_coefficients(expression)
-        # Initialize temporaries and any auxiliary expressions for special
-        # handling
-        self.temps, self.aux_exprs = generate_expr_data(expression)
+        # Initialize temporaries and any auxiliary temporaries
+        temps, aux_exprs, expr_dags = generate_expr_data(expression)
+        # Sort by temporary str: 'T0', 'T1', etc.
+        self.temps = OrderedDict(sorted(iteritems(temps),
+                                        key=lambda x: str(x[1])))
+
+        # Collect the reference count of operands in auxiliary expressions
+        self._ref_counts = collect_reference_count(expr_dags)
+        # Since the most complicated expressions get caught first, we
+        # reorder by expression complexity to create temporaries for the
+        # inner-most instances first (using reference count later to decide
+        # if it's worth the extra memory)
+        self.aux_temps = OrderedDict.fromkeys(sorted(aux_exprs,
+                                                     key=lambda x:
+                                                     x._complexity))
```

Comment:
I wanted to start inside and work my way outside to avoid running into a case where I need to generate an expression for a temp and there are unknowns inside that haven't been declared yet. However, I just removed this and simply used `reversed` (I work my way down an expression by starting from top and move to the roots). Seems to work. Will amend.